### ISSUE TITLE
Encountered internal error running command: Error: zipAlignApk failed

### ISSUE THREAD

**Issue by brunocuratio** [appium/appium#12442]
## The problem

I'm trying to run Appium inspector and I'm getting this error. 

## Environment

* Appium version (or git revision) that exhibits the issue: 1.12.1
* Last Appium version that did not exhibit the issue (if applicable): 1.11
* Desktop OS/version used to run Appium: Sierra 10.13.6
* Node.js version (unless using Appium.app|exe): v11.13.0
* Npm or Yarn package manager: 6.7.0
* Mobile platform/version under test: Android 9
* Real device or emulator/simulator: Device
* Appium CLI or Appium.app|exe: Appium Desktop

## Details

I'm getting this error:

> Encountered internal error running command: Error: zipAlignApk failed. Original error: Command '/Users/curatio/Library/Android/sdk/build-tools/29.0.0-rc1/zipalign -f 4 /Users/curatio/IdeaProjects/ui_tests/apk/Curatio-debug-3.8.0.apk /var/folders/s3/ff6krr157v7g45sl5ysfqrs40000gq/T/201934-64772-19dtlxv.cag5/appium.tmp' exited with code 1. Stdout: ''; Stderr: 'Unable to open '/Users/curatio/IdeaProjects/ui_tests/apk/Curatio-debug-3.8.0.apk' as zip archive 

## Link to Appium logs

https://gist.github.com/brunocuratio/192664b0403c862a6826695062172cbd

**Comment by dpgraham** [comment]
We're in the process of testing SDK version 29 (Android Q) right now. Thanks for bringing this up.

Does this same error happen with SDK 28?

**Comment by brunocuratio** [comment]
I've updated to version 29 and now I downgrade to 28 but appium keep getting this path /Users/curatio/Library/Android/sdk/build-tools/29.0.0-rc1 that don't even exists

> curatio@Bacelar-2 build-tools1 $ cd /Users/curatio/Library/Android/sdk/build-tools/29.0.0-rc1
-bash: cd: /Users/curatio/Library/Android/sdk/build-tools/29.0.0-rc1: No such file or directory
curatio@Bacelar-2 build-tools1 $

**Comment by brunocuratio** [comment]
Having the same issue with sdk 28

> Encountered internal error running command: Error: zipAlignApk failed. Original error: Command '/Users/curatio/Library/Android/sdk/build-tools/28.0.3/zipalign -f 4 /Users/curatio/IdeaProjects/ui_tests/apk/Curatio-debug-3.8.0.apk /var/folders/s3/ff6krr157v7g45sl5ysfqrs40000gq/T/201934-67611-1lff9j3.kde8/appium.tmp' exited with code 1. Stdout: ''; Stderr: 'Unable to open '/Users/curatio/IdeaProjects/ui_tests/apk/Curatio-debug-3.8.0.apk' as zip archive

**Comment by mykola-mokhnach** [comment]
Does it work if the command is executed manually?

**Comment by mykola-mokhnach** [comment]
Perhaps the apk is corrupted or not writeable by appium?

**Comment by matthewsbesthittv** [comment]
I'm also getting the same error:

**Environment**
Appium version (or git revision) that exhibits the issue: 1.8.2-beta
Last Appium version that did not exhibit the issue (if applicable): Unknown
Desktop OS/version used to run Appium: Mojave 10.14.4
Node.js version (unless using Appium.app|exe): v11.13.0
Npm or Yarn package manager: 6.7.0
Mobile platform/version under test: Android u.1
Real device or emulator/simulator: Emulator
Appium CLI or Appium.app|exe: Appium CLI

**Comment by KazuCocoa** [comment]
Did the apk build with `build-tools/29.0.0-rc1`? After removing the 29.0.0, did you re-build the `Curatio-debug-3.8.0.apk` and run with 28? @brunocuratio 

https://github.com/appium/appium/issues/12311#issuecomment-473535496 happened in my env before.

**Comment by brunocuratio** [comment]
@KazuCocoa I'm gonna check with my developer which SDK he used to build.

**Comment by brunocuratio** [comment]
I'm using the same build-tool as my developer. 28.0.2 and still have the same issue

**Comment by KazuCocoa** [comment]
mmm, thanks.
Can you run below command with your apk which built by 28.0.2?
(Just finished building one.)

```
/Users/curatio/Library/Android/sdk/build-tools/28.0.3/zipalign -f 4 your.apk arbitrary/file/appium.tmp
```

If `zipalign`works, `your.apk` will be almost copied into `arbitrary/file/appium.tmp` without errors.
If it works, can you try to run Appium again using the `your.apk`?

**Comment by brunocuratio** [comment]
I had to ask him to build the apk again, now it is working. But it is odd because last week everything as working fine. It seems that after updating to `built tools 29` I've started having this problem.

**Comment by dpgraham** [comment]
29 is still beta so there's probably issues.

**Comment by KazuCocoa** [comment]
I see.
In my case, `zipalign` bundled in the 29 worked if the target apk was built by 28 sdk tools, tho..

You can get rid of 29 tool by deleting `xxxx/build-tools/29.0.0-rc1` directly.
Then, Appium tries to use 28.0.x if the number is the biggest number in `build-tools`.

**Comment by mykola-mokhnach** [comment]
I'm going to add a new feature, so it would be possible to set the target SDK version as capability and use build tools from this particular folder rather than from the most recent one

**Comment by mykola-mokhnach** [comment]
`buildToolsVersion` capability has been added to Android in appium@beta
If one sets this capability to a particular valid build tools version (actually, subfolder of $ANDROID_HOME/build-tools) than Appium will use helper executables from this folder instead of the most recent one.

**Comment by romy21romy** [comment]
well still i am facing this issue. 
can you please provide me information to resolve the issue .

i have just confirm my build tool version is 28.0.3
 thank you

**Comment by majdukovic** [comment]
Thanks @KazuCocoa. I was having same issue and resolved it by deleting `29.0.0-rc1` from `build-tools`.

@romy21romy Delete `29.0.0-rc1` version from `/Users/<username>/Library/Android/sdk/build-tools` (or where your build-tools are), build new apk file and then reopen Appium and try to run it with this new apk version.
